In [28]:
# === Step 1: Import libraries ===
from sentence_transformers import SentenceTransformer, util
import numpy as np
import pandas as pd
import json

with open("../data/referentiel.json", 'r', encoding='utf-8') as f:
    data = json.load(f)

# 2. Création de l'index des compétences
# On itère sur data['blocs'] qui est la liste des blocs
competences_index = {
    comp['id']: comp
    for bloc in data['blocs']
    for comp in bloc['competences']
}

# Test d'accès
print(f"Nombre de compétences indexées : {len(competences_index)}")

Nombre de compétences indexées : 132


In [29]:
id_recherche = "C001"
# .get() est plus sûr que index[id] car il ne fait pas planter le script si l'ID n'existe pas
competence = competences_index.get(id_recherche)

if competence:
    print(f"Competence : {competence}")
    print(f"Texte: {competence['texte']}")
    print(f"Niveau: {competence['niveau']}")

Competence : {'id': 'C001', 'texte': 'Analyser et interpréter des contrats et textes juridiques complexes', 'niveau': 'technique'}
Texte: Analyser et interpréter des contrats et textes juridiques complexes
Niveau: technique


In [30]:
metier_index = {
    metier['id']: metier
    for metier in data['metiers']
    for bloc in metier['blocs_requis']
}

# Test d'accès
print(f"Nombre de compétences indexées : {len(metier_index)}")
print(metier_index)

Nombre de compétences indexées : 40
{'M01': {'id': 'M01', 'titre': "Avocat d'affaires", 'description': 'Conseille les entreprises sur les aspects juridiques et stratégiques de leurs opérations', 'blocs_requis': ['B01', 'B02'], 'competences_cles': ['C001', 'C002', 'C003', 'C009', 'C013', 'C015']}, 'M02': {'id': 'M02', 'titre': "Juriste d'entreprise", 'description': 'Assure la sécurité juridique des opérations commerciales et contractuelles', 'blocs_requis': ['B01', 'B02'], 'competences_cles': ['C001', 'C003', 'C004', 'C005', 'C016', 'C010']}, 'M03': {'id': 'M03', 'titre': 'Notaire', 'description': 'Authentifie les actes juridiques et conseille en gestion patrimoniale', 'blocs_requis': ['B01', 'B04'], 'competences_cles': ['C001', 'C003', 'C008', 'C037', 'C040', 'C010']}, 'M04': {'id': 'M04', 'titre': 'Consultant en conformité RGPD', 'description': 'Assure la conformité des données personnelles et des systèmes informatiques aux réglementations', 'blocs_requis': ['B01', 'B10'], 'competence

In [31]:
id_recherche = "M01"
# .get() est plus sûr que index[id] car il ne fait pas planter le script si l'ID n'existe pas
metier = metier_index.get(id_recherche)

if metier:
    print(f"Competence : {metier}")

Competence : {'id': 'M01', 'titre': "Avocat d'affaires", 'description': 'Conseille les entreprises sur les aspects juridiques et stratégiques de leurs opérations', 'blocs_requis': ['B01', 'B02'], 'competences_cles': ['C001', 'C002', 'C003', 'C009', 'C013', 'C015']}


In [32]:
bloc_index = {
    bloc['id']: bloc
    for bloc in data['blocs']
}

In [33]:
id_recherche = "B08"
# .get() est plus sûr que index[id] car il ne fait pas planter le script si l'ID n'existe pas
metier = bloc_index.get(id_recherche)

if metier:
    print(f"Competence : {metier}")

Competence : {'id': 'B08', 'nom': 'Data Analysis', 'description': 'Compétences en analyse de données, visualisation et statistiques', 'competences': [{'id': 'C085', 'texte': "Programmer en Python ou R pour l'analyse de données", 'niveau': 'technique'}, {'id': 'C086', 'texte': 'Créer des visualisations de données avec Matplotlib, Tableau ou Power BI', 'niveau': 'technique'}, {'id': 'C087', 'texte': 'Appliquer des statistiques descriptives et inférentielles', 'niveau': 'technique'}, {'id': 'C088', 'texte': 'Maîtriser SQL pour interroger et manipuler des bases de données', 'niveau': 'technique'}, {'id': 'C089', 'texte': "Nettoyer et préparer des données brutes pour l'analyse", 'niveau': 'technique'}, {'id': 'C090', 'texte': 'Créer des dashboards interactifs pour le pilotage décisionnel', 'niveau': 'technique'}, {'id': 'C091', 'texte': 'Réaliser des analyses exploratoires de données (EDA)', 'niveau': 'technique'}, {'id': 'C092', 'texte': 'Interpréter les résultats statistiques et communiqu

In [34]:
id_recherche = "M09"
# .get() est plus sûr que index[id] car il ne fait pas planter le script si l'ID n'existe pas
metier = metier_index.get(id_recherche)

if metier:
    print(f"metier : {metier['titre']}")
    print(f"Descritpion : {metier['description']}")
    print(f"id_competence : {metier['competences_cles']}")
    print([[comp['id'], comp['texte']] for cid in metier['competences_cles'] if (comp := competences_index.get(cid))])


metier : Business Analyst
Descritpion : Analyse les données pour éclairer les décisions stratégiques et opérationnelles
id_competence : ['C013', 'C037', 'C041', 'C085', 'C091', 'C093']
[['C013', 'Réaliser une analyse de marché et identifier les opportunités business'], ['C037', 'Réaliser des analyses financières approfondies et diagnostics économiques'], ['C041', 'Créer et analyser des tableaux de bord financiers pour le pilotage'], ['C085', "Programmer en Python ou R pour l'analyse de données"], ['C091', 'Réaliser des analyses exploratoires de données (EDA)'], ['C093', 'Avoir une logique et un raisonnement analytique développés']]


In [35]:
from sentence_transformers import SentenceTransformer
import json
import numpy as np

# 1. Charger le modèle SBERT (all-MiniLM-L6-v2 est recommandé dans ton sujet)
# C'est un modèle léger, rapide et performant en local [cite: 323]
model = SentenceTransformer('all-MiniLM-L6-v2')
# 4. Générer les embeddings


In [36]:
textes = [c['texte'] for c in competences_index.values()]
embeddings = model.encode(textes, convert_to_tensor=True)

# 5. Stocker les résultats pour le moteur de recherche
# On associe chaque ID à son vecteur
vector_store = {
    comp['id']: embeddings[i]
    for i, comp in enumerate(competences_index.values())
}

print(f"Référentiel vectorisé : {len(vector_store)} compétences prêtes.")

Référentiel vectorisé : 132 compétences prêtes.


In [37]:
from sentence_transformers import util

def calculer_similarites(cheminentree_utilisateur, vector_store, model):
    """
    Compare une saisie utilisateur avec toutes les compétences indexées.
    """
    # 1. Transformer la saisie utilisateur en vecteur (Embedding)
    user_embedding = model.encode(entree_utilisateur, convert_to_tensor=True)

    scores = []

    # 2. Comparer avec chaque compétence du référentiel
    for comp_id, comp_embedding in vector_store.items():
        # Calcul de la similarité cosinus (résultat entre 0 et 1)
        score = util.cos_sim(user_embedding, comp_embedding).item()

        scores.append({
            "id": comp_id,
            "score": round(score, 4)
        })

    # 3. Trier par score décroissant pour avoir les meilleures correspondances
    scores = sorted(scores, key=lambda x: x['score'], reverse=True)
    return scores

# --- TEST ---
input_test = "Je sais piloter des campagnes publicitaires sur Facebook Ads et optimiser le référencement naturel des sites web."
resultats = calculer_similarites(input_test, vector_store, model)

print(f"Top 20 correspondances pour : '{input_test}'")
text_cross = [competences_index[res['id']]['texte'] for res in resultats[:20]]
for res in resultats[:20]:
    # On utilise l'index de compétences créé précédemment pour afficher le texte
    texte = competences_index[res['id']]['texte']
    print(f"-{competences_index[res['id']]['id']} {texte} (Score: {res['score']})")

Top 20 correspondances pour : 'Je sais piloter des campagnes publicitaires sur Facebook Ads et optimiser le référencement naturel des sites web.'
-C079 Gérer des campagnes de publicité en ligne (Facebook Ads, Google Ads) (Score: 0.7709)
-C073 Gérer des communautés en ligne et animer les réseaux sociaux (Score: 0.5513)
-C078 Élaborer et déployer une stratégie social media multi-plateformes (Score: 0.4893)
-C031 Prospecter de nouveaux clients et développer un portefeuille commercial (Score: 0.4774)
-C034 Faire preuve de créativité dans les approches commerciales (Score: 0.4757)
-C063 Développer des créations publicitaires impactantes et mémorables (Score: 0.4755)
-C090 Créer des dashboards interactifs pour le pilotage décisionnel (Score: 0.4614)
-C025 Maîtriser les techniques de marketing digital et e-commerce (Score: 0.4574)
-C074 Créer du contenu digital engageant (vidéos, posts, stories) (Score: 0.4565)
-C119 Aimer construire des systèmes et voir ses créations fonctionner (Score: 0.45

In [38]:
from sentence_transformers import CrossEncoder

# 1. Load a pretrained CrossEncoder model
model_cross = CrossEncoder("cross-encoder/ms-marco-MiniLM-L6-v2")
# The texts for which to predict similarity scores
query = input_test
passages = text_cross

# 2a. Either predict scores pairs of texts
scores = model_cross.predict([(query, passage) for passage in passages])
#print(scores)
# => [8.607139 5.506266 6.352977]

# 2b. Or rank a list of passages for a query
ranks = model_cross.rank(query, passages, return_documents=True)

import math

def sigmoid(x):
    return 1 / (1 + math.exp(-x))
# Exemple pour ton top score (2.22)

print("Query:", query)
for rank in ranks[:3]:
    print(f"- #{rank['corpus_id']} ({rank['score']:.2f}): {rank['text']}")


Query: Je sais piloter des campagnes publicitaires sur Facebook Ads et optimiser le référencement naturel des sites web.
- #0 (1.72): Gérer des campagnes de publicité en ligne (Facebook Ads, Google Ads)
- #15 (-0.73): Optimiser le référencement naturel SEO et payant SEA
- #5 (-4.41): Développer des créations publicitaires impactantes et mémorables


In [39]:
print(resultats[1])

{'id': 'C073', 'score': 0.5513}


Resultat du test
Query: J'ai une solide expérience dans la rédaction de clauses contractuelles et le suivi des contentieux commerciaux.
- #0 (-5.06): Posséder une excellente qualité rédactionnelle et syntaxe
- #1 (-7.09): Gérer les procédures juridiques et le contentieux
- #6 (-8.65): Avoir une grande aisance à l'oral et en présentation publique

Query: Je développe des scripts Python pour nettoyer des bases SQL et entraîner des modèles de classification.
- #1 (-1.35): Maîtriser SQL pour interroger et manipuler des bases de données
- #9 (-2.42): Nettoyer et préparer des données brutes pour l'analyse
- #4 (-3.64): Développer et entraîner des algorithmes de machine learning supervisé et non supervisé

Query: Je sais piloter des campagnes publicitaires sur Facebook Ads et optimiser le référencement naturel des sites web.
- #0 (1.72): Gérer des campagnes de publicité en ligne (Facebook Ads, Google Ads)
- #15 (-0.73): Optimiser le référencement naturel SEO et payant SEA
- #5 (-4.41): Développer des créations publicitaires impactantes et mémorables

